File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0900_RLE_Iterator.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0900_RLE_Iterator {
    private int index;
    private int[] encoding;

    public Question_0900_RLE_Iterator(int[] encoding) {
        this.encoding = encoding;
        this.index = 0;
    }

    public int next(int n) {
        while (index < encoding.length && n > encoding[index]) {
            n = n - encoding[index];
            index += 2;
        }
        if (index >= encoding.length) {
            return -1;
        }
        encoding[index] = encoding[index] - n;
        return encoding[index + 1];
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0900_RLE_Iterator.cc
#include <vector>

using namespace std;

class RLEIterator {
 public:
  RLEIterator(vector<int>& encoding): vec(encoding), index(0) {
  }

  int next(int n) {
    while (index < vec.size() && n > vec[index]) {
      n = n - vec[index];
      index += 2;
    }
    if (index >= vec.size()) {
      return -1;
    }
    vec[index] = vec[index] - n;
    return vec[index + 1];
  }

 private:
  vector<int> vec;
  int index;
};

